File: discipline_client_ts/chronic/monotonic_clock.ts
import { Duration, DateTime, Instant, Branded } from "../x.ts";

const BRAND = Symbol();

export type MonotonicClock = Branded<typeof BRAND, {
  elapsedTime: Duration,
  previousSynchronizationTime: DateTime,
}>;

export const construct = (
  elapsedTime: Duration,
  previousSynchronizationTime: DateTime,
): MonotonicClock => {
  return Branded(BRAND, {
    elapsedTime,
    previousSynchronizationTime,
  });
};

export const create = (now: DateTime): MonotonicClock => {
  return construct(
    Duration.zero(),
    now,
  );
};

export const synchronize = (it: MonotonicClock, now: DateTime): void => {
  const interval = DateTime.tillOrZero(it.previousSynchronizationTime, now);
  it.elapsedTime = Duration.plusOrMax(it.elapsedTime, interval);
  it.previousSynchronizationTime = now;
};

export const getElapsedTime = (it: MonotonicClock): Duration => {
  return it.elapsedTime;
};

export const getPreviousSynchronizationTime = (it: MonotonicClock): DateTime => {
  return it.previousSynchronizationTime;
};

export const getNow = (it: MonotonicClock): Instant => {
  return Instant.fromElapsedTime(it.elapsedTime);
};

export const getNowAsDateTime = (it: MonotonicClock): DateTime => {
  const now = DateTime.now();
  if (DateTime.isEarilerThan(now, it.previousSynchronizationTime)) {
    return now;
  } else {
    return it.previousSynchronizationTime;
  }
};

export const MonotonicClock = {
  create,
  construct,
  synchronize,
  getElapsedTime,
  getPreviousSynchronizationTime,
  getNow,
  getNowAsDateTime,
};

File: discipline4/app/src/main/java/com/example/app/chronic/MonotonicClock.kt
package com.example.app

import com.example.app.*
import androidx.room.Entity

/**
 * A monotonic clock that only moves forward
 */
@Entity
public data class MonotonicClock(
  var elapsedTime: Duration,
  var previousSynchronizationTime: Instant,
  var synchronizationInterval: Duration,
) {
  companion object {
    fun create(synchronizationInterval: Duration): MonotonicClock {
      return MonotonicClock(
        elapsedTime = Duration.zero(), 
        previousSynchronizationTime = Instant.fromElapsedTime(Duration.zero()),
        synchronizationInterval,
      )
    }

    fun constructOrThrow(
      elapsedTime: Duration,
      previousSynchronizationTime: Instant,
      synchronizationInterval: Duration,
    ): MonotonicClock {
      // TODO: Vallidate and throw
      return MonotonicClock(
        elapsedTime = elapsedTime,
        previousSynchronizationTime = previousSynchronizationTime,
        synchronizationInterval = synchronizationInterval,
      )
    }
  }
  
  /**
   * Synchronizes the clock, adding elapsed time since last sync
   */
  fun synchronize(now: Instant) {
    val interval = previousSynchronizationTime.tillOrZero(now)
    elapsedTime = elapsedTime.plusOrMax(interval)
    previousSynchronizationTime = now
  }
  
  fun getElapsedTime(): Duration {
    return elapsedTime
  }

  fun getPreviousSynchronizationTime(): Instant {
    return previousSynchronizationTime
  }
  
  /**
   * Returns the current instant based on elapsed time
   */
  fun getNow(): Instant {
    return Instant.fromElapsedTime(elapsedTime)
  }
  
  // /**
  //  * Returns the current DateTime (clamped to not exceed last sync time)
  //  */
  // fun getNowAsDateTime(): DateTime {
  //   val now = DateTime.now()
  //   return if (now.isEarlierThan(previousSynchronizationTime)) {
  //     now
  //   } else {
  //     previousSynchronizationTime
  //   }
  // }
}